module/os/linux/zfs/zfs_znode.c

```diff
@@ -1111,6 +1111,10 @@ zfs_zget(zfsvfs_t *zfsvfs, uint64_t obj_num, znode_t **zpp)
 			mutex_exit(&zp->z_lock);
 			sa_buf_rele(db, NULL);
 			zfs_znode_hold_exit(zfsvfs, zh);
+			/* if znode is already marked for deletion, break out early */
+			if (zp->z_unlinked) {
+				return (SET_ERROR(ENOENT));
+			}
```

Comment:
The fix looks good, I'd just suggest we restructure this block a little bit.  What do you think of this (untested), it has a couple small advantages:

* This allows us to skip the `igrab()` entirely when `zp->z_unlinked` is set.
* There's a single `mutex_exit() / sa_buf_rele() / ZFS_OBJ_HOLD_EXIT()` path.
* It aligns this logic a little more closely with the [FreeBSD](https://github.com/freebsd/freebsd/blob/master/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/zfs_znode.c#L1194L1207) and [illumos](https://github.com/illumos/illumos-gate/blob/master/usr/src/uts/common/fs/zfs/zfs_znode.c#L1212L1224) versions of this function (if you squint).  All of the platforms need this same unlinked check, but the VFS integration specifics are different on each platform.

```c
                mutex_enter(&zp->z_lock);
                ASSERT3U(zp->z_id, ==, obj_num);
                /*
                 * If zp->z_unlinked is set then the znode is marked
                 * for deletion and should not be discoverable.
                 *
                 * If igrab() returns NULL the VFS has independently
                 * determined the inode should be evicted and has
                 * called iput_final() to start the eviction process.
                 * The SA handle is still valid but because the VFS
                 * requires that the eviction succeed we must drop
                 * our locks and references to allow the eviction to
                 * complete.  The zfs_zget() may then be retried.
                 */
                if (zp->z_unlinked) {
                        err = SET_ERROR(ENOENT);
                } else if (igrab(ZTOI(zp)) == NULL) {
                        err = SET_ERROR(EAGAIN);
                        cond_resched();
                } else {
                        *zpp = zp;
                        err = 0;
                }

                mutex_exit(&zp->z_lock);
                sa_buf_rele(db, NULL);
                zfs_znode_hold_exit(zfsvfs, zh);

                if (err == EAGAIN)
                        goto again;

                return (err);
```

Reply:
The fix looks good to me too, but I like the revised version better. I suggest we go with that and add an original-patch-by line to the commit.

Reply:
I agree that we should have the `if (zp->z_unlinked) err = ENOENT;` code, like illumos.  @behlendorf's way of doing that seems cleanest to me.

Reply:
Thank you for the review! I think we should go with this version as well, it looks much cleaner and has the mentioned advantages over the original patch. Apologies for the initial "naive" fix, I'm still finding my way around the ZFS code and didn't stop to think and look at the illumos code...

I can respin the PR to be more aligned with this one, or we can go with @ryao's original-patch-by suggestion. Either one is fine by me, so please let me know how you would like to proceed!

Reply:
@hrasiq

Great.  We're regression testing this, and will double-check with the user/reporter specific workload that it's all good.  Thanks!

Reply:
@hrasiq How about: Who ever gets it in for review first?
It's not worth the time debating who is sending it in.

And don't worry about it, it's not a naive fix you did... People having feedback doesn't mean your code is bad or naive ;)

@mfoliveira What are you regression testing? The proposed changes in this Review or the original fix?

Reply:
@Ornias1993 The proposed changes (reported to be untested). The original fix has been regression tested for the submission.

Reply:
@Ornias1993 In that case, allow me to re-submit this one :)